File: packages/engine/src/modules/excavation.ts
import { PHYSICAL_CONSTANTS } from "../constants";
import type { SimParams } from "../types";

export interface ExcavationOutput {
  cuttingForceN: number;
  mechPowerW: number;
  fleetMassKg: number;
  secExcavation_JPerKg: number;
  regolithPerKgProduct: number;
}

export function regolithPerKgProduct(params: SimParams, xO2Effective?: number): number {
  if (params.site === "equatorial") {
    return 1 / (xO2Effective ?? params.xO2 * params.fExtract);
  }
  return 1 / params.chiIce;
}

export function simulateExcavation(params: SimParams, xO2Effective?: number): ExcavationOutput {
  const gL = PHYSICAL_CONSTANTS.gL.value;
  const q = params.rhoReg * gL * params.zDepth;
  const cuttingForceN =
    (params.c * params.Nc +
      q * params.Nq +
      0.5 * params.rhoReg * gL * params.wBlade * params.dBlade * params.Ngamma) *
    params.wBlade *
    params.dBlade;
  const mechPowerW = (cuttingForceN * params.vCut) / params.etaDrive;
  const regolithPerKg = regolithPerKgProduct(params, xO2Effective);
  const secExcavation_JPerKg = params.eMining * regolithPerKg;
  const fleetMassKg = params.kExcFleet * params.targetKgPerDay;

  return {
    cuttingForceN,
    mechPowerW,
    fleetMassKg,
    secExcavation_JPerKg,
    regolithPerKgProduct: regolithPerKg
  };
}

File: python/selene_isru/modules/excavation.py
from __future__ import annotations

from typing import Any

from ..constants import c


def regolith_per_kg_product(params: dict[str, Any], x_o2_effective: float | None = None) -> float:
    if params["site"] == "equatorial":
        return 1 / (x_o2_effective if x_o2_effective is not None else params["xO2"] * params["fExtract"])
    return 1 / params["chiIce"]


def simulate_excavation(params: dict[str, Any], x_o2_effective: float | None = None) -> dict[str, float]:
    g_l = c("gL")
    q = params["rhoReg"] * g_l * params["zDepth"]
    cutting_force_n = (
        params["c"] * params["Nc"]
        + q * params["Nq"]
        + 0.5 * params["rhoReg"] * g_l * params["wBlade"] * params["dBlade"] * params["Ngamma"]
    ) * params["wBlade"] * params["dBlade"]
    mech_power_w = cutting_force_n * params["vCut"] / params["etaDrive"]
    regolith_per_kg = regolith_per_kg_product(params, x_o2_effective)
    sec_excavation_j_per_kg = params["eMining"] * regolith_per_kg
    fleet_mass_kg = params["kExcFleet"] * params["targetKgPerDay"]

    return {
        "cuttingForceN": cutting_force_n,
        "mechPowerW": mech_power_w,
        "fleetMassKg": fleet_mass_kg,
        "secExcavation_JPerKg": sec_excavation_j_per_kg,
        "regolithPerKgProduct": regolith_per_kg,
    }
